### ISSUE TITLE
Maximum system size

### ISSUE THREAD

**Issue by dihm** [jrenaud90/CyRK#74]
So I finally got around to updating my older `cyrk_ode` code to use `pysolver_ivp`, which was a very smooth process. But then I tried one of our larger test systems and discovered the new C++ backend has a different limit for the system size (32 instead of 65535). We somewhat regularly hit the old ushort limit, so dropping to 32 makes things fairly challenging. Of course, I strongly suspect there was a good reason to change the size, so I'm curious to know what appetite you have to increase that limit, and what it could be increased to? Or perhaps if the limit could be configurable at run-time without significant detriment to others with more sane system sizes?

To establish a sense of scale, we are solving quantum master equations, so the system size is `n^2-1`, where `n` is the number of states in the problem. So the scaling is abysmal. Common, simple systems (for us) have `n=4-10`, giving eqs 15-99. A system of 16 states gives 255 eqs. And we envision doing systems on the order of `n>=50` very soon (so 2499 eqs).

**Comment by jrenaud90** [comment]
wow that is a lot of differential equations!

So I have been debating what to do about this. Right now, CyRK compiles some arrays (e.g., dy; y; and arrays that depend on their size) to the stack to improve performance and avoid expensive heap allocations. However, to stack allocate I have to know the number of variables at compile time. 32 is usually overkill for my types of problems so typically I am underutilizing these allocations. Whatever choice I make, it will be set for all users of CyRK so folks who are only doing a ODE with 4 equations would be stuck with memory usage (and related cache misses) for a 65,000 ODE (it is a lot worse than this because there are several arrays that need to be allocated with this size).

The answer is to heap allocate these arrays. Let me do some testing to see how much that actually impacts performance. If it is not too bad then it will be easy to implement (will likely put it into v0.12.0 since it could affect folks' code).

**Workaround**
In the mean time, if you want to get around this limit you can do:
- ```git clone https://github.com/jrenaud90/CyRK``` to a preferred directory. 
- Edit the file CyRK/CyRK/cy/common.hpp 
  - and change:
  - "Y_LIMIT  = 65535"
  - and 
  - "DY_LIMIT    = 65535" (or larger if you are capturing [extra outputs](https://github.com/jrenaud90/CyRK/blob/main/Documentation/Extra%20Output.md))
- Go back to top level directory and edit the "pyproject.toml" file and change  "version = '0.11.3'" to something like "version = '0.11.4a0'" just so you make sure you are using the right version.
- Open a terminal and manually install:
  - `python -m pip uninstall cyrk` (get rid of old ones)
  - `python -m pip install -v .`

Let me know if that works for you (also would be curious how it affects performance).

By the way the newest version of CyRK, v0.11.3, has several performance improvements and bugfixes so I would go ahead an work off of that version from now on!

Also it is on conda now ```conda install -c conda-forge  cyrk``` (but you will likely want to do the git clone method to make your edits).

**Comment by jrenaud90** [comment]
Hey @dihm, I implemented a new backend that uses heap allocated arrays and removes the hard limit on dependent variables. As expected, I am noticing a performance hit but its pretty minor, particularly for `pysolve_ivp`. I am going to think about ways to optimize for the cache but I think it makes sense to have CyRK switch to this method to remove that limitation.

You can try it out for yourself if you git clone the repository, and `git checkout Experimenting-with-heap-allocated-arrays` then install locally with `python -m pip install -v .` from a terminal in the directory w/ setup.py.

**Comment by dihm** [comment]
Thanks @jrenaud90! That change gives me the same performance I had with the old `cyrk_ode`. For a problem that works in either case (15 eqs) I see little difference either way (potentially a small slow-down, 4s -> 4.2s going from stack to heap). Not sure how much I'd trust my profiling though as my use case requires dense output via `t_eval` which complicates things, not to mention it has some dependence on the numba caching of the inputs that seems to be a little sporadic.

Regardless, if you can tolerate the slowdown for your purposes, it is a great help to me.

**Comment by jrenaud90** [comment]
Hey @dihm - This has been implemented in v0.11.4. Let me know if you run into any issues.

**Comment by jrenaud90** [comment]
Oh no! That is odd that it is happening with such a specific number. If you could make an example that replicates the crash then I can take a look at it.

**Comment by dihm** [comment]
Hmm. My simple plan to solve a bunch of simple diffeqs in parallel has not triggered the issue (though I may have found a case where the result success reports False but the message and actual output appear to have worked correctly, with Neqs>2**21).

I am now more strongly suspicious of the code on my end being the likely culprit. I'll have to do some more debugging to see if I can track down what the issue is.

**Comment by jrenaud90** [comment]
Well either way the crash should not happen. If there is a true error with your input then cyrk should raise an exception rather than just crash. So something is still fishy on cyrks end. So if you do find a way to reproduce it please feel free to post so I can dig into what's happening.

**Comment by jrenaud90** [comment]
Discussing the new issue over in issue #78. Closing this one.
